File: py/PigLatin.py
"""
https://www.codewars.com/kata/520b9d2ad5c005041100000f/solutions/java
https://www.codewars.com/kata/520b9d2ad5c005041100000f/train/java
https://www.codewars.com/kata/520b9d2ad5c005041100000f/train/python

RESULTS:
$ python PigLatin.py 
PASS |elloHay orldway !|   ---> |elloHay orldway !|
PASS |igPay atinlay siay oolcay|   ---> |igPay atinlay siay oolcay|
PASS |igPay atinlay.|   ---> |igPay atinlay.|
PASS |zyay!?!|   ---> |zyay!?!|
PASS |zyay !|   ---> |zyay !|
FAIL |o|   ---> |oay|
PASS |.|   ---> |.|

"""
import re
from Tester import verdict, log

# public static String java_version_of_pig_it(String original) {
#        return original.replaceAll("(\\w)(\\w*)", "$2$1ay");
# }

# loren
def pig_it(sentence):
    # Pretty code!

    return re.sub(r"\b(\w)(\w+)", r"\2\1ay",sentence)

# mine follows
def pig_it_mine(original):
    # code ugly, but does pass all the tests
    words = list(original.split(" ")) 
    x = " ".join(map(pigify, words))
    return x 

def pigify(word):
    latin = ""
    punctuation = ""  # maybe a . Maybe a !! Maybe a  !?
    result = ""
    for i in range(len(word)):
        c = word[i]
        if i == 0:
            if c.isalnum():
                latin = c + "ay"
            else:
                punctuation = c
        else:
            if c.isalnum():
                result += c
            else:
                punctuation += c 
    return result + latin + punctuation

if __name__ == '__main__':

        verdict(pig_it("Hello world !"),"elloHay orldway !")
        verdict(pig_it("Pig latin is cool"), "igPay atinlay siay oolcay")
        verdict(pig_it("Pig latin."), "igPay atinlay.")
        verdict(pig_it("yz!?!"), "zyay!?!") # no space + 3 weirds
        verdict(pig_it("yz !"), "zyay !") # space before !
        verdict(pig_it("o"), "oay") # SIngle characters
        verdict(pig_it("."), ".") # SIngle period
        

File: java/PigLatin.java
//https://www.codewars.com/kata/520b9d2ad5c005041100000f/train/java
/*
pigIt('Pig latin is cool'); // igPay atinlay siay oolcay
pigIt('Hello world !');     // elloHay orldway !
*/

import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.Arrays;
import java.util.stream.Collectors;

public class PigLatin extends Tester {

    private static String logic(String original ) {
        char[] charArray = original.toCharArray();
        int i = 0; 
        String latin = ""; 
        String punctuation = "";
        String result = ""; 
        for(char c:charArray) {
            if ( i == 0 ) {
                if (Character.isLetterOrDigit(c)) {
                    latin = String.valueOf(c) + "ay";
                } else {
                    punctuation  += String.valueOf(c);
                }
            } else {
                if (Character.isLetterOrDigit(c)) {
                    result += String.valueOf(c); 
                } else {
                    punctuation  += String.valueOf(c);
                }
            }
            i++;
        }
        return result + latin + punctuation;
    }

	public static String pigIt_mine(String original) {
        String space = " "; 
        return Stream
            .of(original.split(space))
            .map(word -> logic(word))
            .collect(Collectors.joining(space));

	}
    public static String pigIt(String original) {
        return original.replaceAll("(\\w)(\\w*)", "$2$1ay");

    }

    public static String pigIt_second(String str) {
        return Arrays.stream(str.split(" ")).map(PigLatin::pigify).collect(Collectors.joining(" "));
    }

    private static String pigify(String word){
        return word.length() > 1 || Character.isLetter(word.charAt(0)) ? word.substring(1) + word.charAt(0) + "ay" : word;
               
    }

	public static void main(String... strings) {
        verdict(pigIt("Hello world !"),"elloHay orldway !");
        verdict(pigIt("Pig latin is cool"), "igPay atinlay siay oolcay");
        verdict(pigIt("Pig latin."), "igPay atinlay.");
        verdict(pigIt("z!"), "zay!");
        verdict(pigIt("z!"), "zay!");
        verdict(pigIt("yz!"), "zyay!");
        verdict(pigIt("yz !"), "zyay !");
        verdict(pigIt("o"), "oay");
        verdict(pigIt("."), ".");

	}
}


